### PR TITLE
Implement modularity in the form of subexpressions

### PR DESCRIPTION
**Pull request by francisrstokes**
## Implement modularity in the form of subexpressions

This PR addresses issue #12. It was a little complex to implement, because there are some edge cases to look out for when merging one SuperExpressive instance with another. These came primarily in the form of dealing with flags, start/end of input markers, named capture group naming collisions, backreference indexing, and deep levels of nesting.

I've added a set of tests which I think cover these edge cases acceptably.

- [ ] This PR  only introduces changes to documentation.
  - Please include a summary of changes and an explanation
- [x] This PR adds new functionality
  - [x] Is there a related issue?
    - Please note the related issue below, and how this PR relates to the issue if appropriate
  - [x] Does the code style reasonably match the existing code?
  - [x] Are the changes tested (using the existing format, as far as is possible?)
  - [x] Are the changes documented in the readme with a suitable example?
  - [x] Is the table of contents updated?
  - [x] Is the `index.d.ts` file updated, using appropriate types and using the same description as the readme?
- [ ] This PR introduces some other kind of change
  - Please explain the change below

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -44,22 +44,24 @@ const quantifierTable = {
   betweenLazy: metadata => `{${metadata[0]},${metadata[1]}}?`,
 }
 
+const applySubexpressionDefaults = expr => {
+  const out = { ...expr };
+  out.namespace = ('namespace' in out) ? out.namespace : '';
+  out.ignoreFlags = ('ignoreFlags' in out) ? out.ignoreFlags : false;
+  out.ignoreStartAndEnd = ('ignoreStartAndEnd' in out) ? out.ignoreStartAndEnd : false;
```
Comment (jcao219) [inline index.js:51]: I wonder if these should default to `true`.

To me, flags are a global concept which sort of clashes with subexpressions.

Comment (francisrstokes) [reply index.js:51]: Yeah I think I agree actually. Explicitly granting the permission seems like it would cause less unexpected problems for people in the end.

File: index.js [hunk 2]
```diff
@@ -441,6 +462,122 @@ class SuperExpressive {
     return next;
   }
 
+  static [mergeSubexpression](el, options, parent, incrementCaptureGroups) {
+    let nextEl = deepCopy(el);
+
+    if (nextEl.type === 'backreference') {
+      nextEl.metadata += parent.state.totalCaptureGroups;
+      incrementCaptureGroups();
```
Comment (francisrstokes) [inline index.js:470]: This isn't right - a backreference shouldn't increase the total number of capture groups. Need an additional check below for `capture` which would increase the count.

File: index.js [hunk 3]
```diff
@@ -441,6 +462,122 @@ class SuperExpressive {
...
+
+      parent.state.hasDefinedEnd = true;
+    }
+
+    return nextEl;
+  }
+
+  subexpression(expr, opts = {}) {
+    assert(expr instanceof SuperExpressive, `expr must be a SuperExpressive instance`);
+    assert(
```
Comment (francisrstokes) [inline index.js:545]: Maybe add a test case for this.

File: index.js [hunk 4]
```diff
@@ -517,6 +655,11 @@ class SuperExpressive {
       case 'namedBackreference': return `\\k<${el.metadata}>`;
       case 'backreference': return `\\${el.metadata}`;
 
+      case 'subexpression': {
```
Comment (francisrstokes) [inline index.js:658]: Might be better to mark `t.subexpression` with `quantifierRequiresGroup`, and let any quantifiers deal with adding the group.
